components/script/dom/document.rs

```diff
@@ -4082,6 +4090,31 @@ impl Document {
     pub(crate) fn set_declarative_refresh(&self, refresh: DeclarativeRefresh) {
         *self.declarative_refresh.borrow_mut() = Some(refresh);
     }
+
+    /// <https://html.spec.whatwg.org/multipage/interaction.html#visibility-state>
+    fn update_visibility_state(&self, visibility_state: DocumentVisibilityState) {
+        // Step 1
+        if self.visibility_state.get() == visibility_state {
+            return;
+        }
+        // Step 2
+        self.visibility_state.set(visibility_state);
+        // Step 3 TODO get timestamp
+        let entry = VisibilityStateEntry::new(
+            &self.global(),
+            visibility_state,
+            *self.global().performance().Now(),
+        );
+        self.window
+            .Performance()
+            .queue_entry(entry.upcast::<PerformanceEntry>());
+        // Step 4 TODO ScreenOrientation hasn't implemented yet
+        // Step 5 TODO ScreenOrientation hasn't implemented yet
+        // Step 6 ???
```

Comment:
@gterzian I opened a draft for now because I want to know if I can skip this.
It seems Servo doesn't have the implementation in steps 4 & 5, and step 6 is so vague that it even has a comment discussing it.